README
NAME
    CatalystX::Controller::Tabs - Automatically build tab sets for a
    controller

SYNOPSIS
      package MyApp::Controller::Something;
      use base qw( Catalyst::Controller CatalystX::Controller::Tabs );
      
  sub friends : Local Tab('Friends') {
        my ( $self, $c ) = @_;
        
    $c->stash->{ 'template' } = 'friends_list.tt2';
      }
      
  sub enemies : Local Tab('Enemies') {
        my ( $self, $c ) = @_;
        
    $c->stash->{ 'template' } = 'enemies_list.tt2';
      }
      
  sub auto : Private {
        my ( $self, $c ) = @_;
        
    $c->forward( 'build_tabs' )
      }

      [%# Template friends_list.tt2 %]
      [% PROCESS tabs.tt2 %]
      <ul>[% FOR enemy IN enemies %]<li>[% enemy %]</li>[% END %]</ul>
      
  [%# Template enemeies_list.tt2 %]
      [% PROCESS tabs.tt2 %]
      <ul>[% FOR friend IN friends %]<li>[% friend %]</li>[% END %]</ul>
      
  [%# Template friends_list.tt2 %]
      <ul>
      [% FOR tab IN tabs %]
          <li[% IF tab.selected %] class="selected">
              <a href="[% tab.uri %]">[% tab.label %]</a>
          </li>
      [% END %]
      </ul>

DESCRIPTION
    This module allows you to add a 'Tab' attribute to action endpoints, and
    it will automatically build a data structure suitable for rendering
    'tabs' to switch between the methods that share the same tab structure.

    Although this was originally built to help with making tabbed
    interfaces, it isn't limited to creating tabs, as it simply collects the
    information about the related actions. Actions are considered to be
    related if they share a namespace.

METHODS
  build_tabs
    This is the only method provided by this module, and you need to make
    sure you arrange to forward to this method when you want the tab data
    structure to be built.

CONFIGURATION
    You can configure this module with the normal Catalyst style of
    controller configuration. The default configuration is:

        __PACKAGE__->config( {
            tabs    => {
                stash_key       => 'tabs',
            }
        } );

  CONFIGURATION OPTIONS
    stash_key
        This configuration option controls the key that will be used to
        store the the generated data structure in the stash.

SAMPLE CSS
    Here is some CSS that works with the template included in the synopsis.
    It's probably not exactly what you need, but it should give a decent
    starting point...

      ul.tabs {
        text-align: left;
        margin: 1em 0 1em 0;
        border-bottom: 1px solid #6c6;
        list-style-type: none;
        padding: 3px 10px 3px 10px;
      }
          
  ul.tabs li {
        display: inline;
      }
      
  ul.tabs li.selected {
        border-bottom: 1px solid #fff;
        background-color: #fff;
      }
      
  ul.tabs li.selected a {
        background-color: #fff;
        color: #000;
        position: relative;
        top: 1px;
        padding-top: 4px;
      }
      
  ul.tabs li a {
        padding: 3px 4px;
        border: 1px solid #6c6;
        background-color: #cfc;
        color: #666;
        margin-right: 0px;
        text-decoration: none;
        border-bottom: none;
      }
      
  ul.tabs a:hover {
        background: #fff;
    }

SEE ALSO
    <http://www.jasonkohles.com/software/CatalystX-Controller-Tabs>

    <http://catalyst.perl.org/>

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc CatalystX::Controller::Tabs

    You can also look for information at:

    *   Project home page

        <http://www.jasonkohles.com/software/CatalystX-Controller-Tabs>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/CatalystX-Controller-Tabs>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/CatalystX-Controller-Tabs>

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=CatalystX-Controller-Tabs>

    *   Search CPAN

        <http://search.cpan.org/dist/CatalystX-Controller-Tabs>

BUGS
    Please report any bugs or feature requests to
    "bug-catalystx-controller-tabs at rt.cpan.org", or through the web
    interface at <http://rt.cpan.org>. I will be notified, and then you'll
    automatically be notified of progress on your bug as I make changes.

  KNOWN BUGS
    The methods provided for determining the URI for an action don't seem to
    work the same way for Catalyst::DispatchType::Regex actions, but this
    hasn't been a big issue for me so far, and I haven't had time to track
    the problem down yet (it seems to be an issue with
    Catalyst::DispatchType::Regex/uri_for_action if you want to look for
    it...)

AUTHOR
    Jason Kohles "<email@jasonkohles.com>" <http://www.jasonkohles.com>

COPYRIGHT AND LICENSE
    Copyright (C) 2007-2008 Jason Kohles.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension CatalystX-Controller-Tabs

0.02  Wed Jan 30 10:17:46 EST 2008
    - Added 'ok' to the build_requires, oops.

0.01  Tue Jan 29 20:54:32 EST 2008
    - initial release


t/Makefile.PL
# This is just here to make sure that Catalyst can find the right location for
# TestApp


t/test-lib.pl
#!/usr/bin/perl -w
##################
# Copyright Jason Kohles (JSK), <email@jasonkohles.com>
# $Id$
##################
use strict;
use warnings;
use FindBin qw( $Bin );
use lib "$Bin/lib";
eval "use TestApp";
if ( $@ ) { plan skip_all => "Couldn't load TestApp: $@"; exit; }
use Test::WWW::Mechanize::Catalyst 'TestApp';
use Data::Dump qw( dump );
use Carp qw( croak verbose );
use Test::More;
use Scalar::Util qw( blessed );

{
    my $mech = Test::WWW::Mechanize::Catalyst->new;
    sub mech { $mech }
}
{
    my $numtabs;
    sub numtabs {
        if ( @_ ) { $numtabs = shift }
        return $numtabs;
    }
}
{
    my $tabs;
    sub settabs { $tabs = ref $_[0] eq 'ARRAY' ? $_[0] : \@_ }
    sub tabs { return wantarray ? @{ $tabs } : $tabs }
    sub tab { return $tabs->[ shift() - 1] }
}

sub matches_ok {
    my ( $item, $format, @vals ) = @_;
    my @check = map { sprintf( $format, $_ ) } 1 .. numtabs();
    if ( ! @vals ) { @vals = map { $_->{ $item } } tabs() }
    is_deeply( \@vals, \@check, "$item matches ok" );
}
sub list_ok {
    my ( $item, @list ) = @_;
    my @vals = map { $_->{ $item } } tabs();
    is_deeply( \@vals, \@list, "$item list matches ok" );
}

sub test_tabs {
    numtabs( shift() );
    get( shift() );
    my %opts = @_;

    for my $key ( sort keys %opts ) {
        my $exp = $opts{ $key };
        my @objects = tabs();

        my @expected = ( $exp ) x scalar @objects;
        if ( ! ref( $exp ) ) {
            if ( $exp =~ /%d/ ) {
                @expected = map { sprintf( $exp, $_ ) } 1 .. @objects;
            } elsif ( $exp =~ /\%\{(\w+)\}/ ) {
                my $x = $1;
                for my $i ( 0 .. $#expected ) {
                    my $val = get_value( $x => $objects[ $i ] );
                    $expected[ $i ] =~ s/\%\{$x\}/$val/g;
                }
            }
        } elsif ( ref $exp eq 'ARRAY' ) {
            @expected = @{ $exp };
        }

        my $sub;
        if ( $key =~ s/^(\w+)\.// ) {
            $sub = $1;
            @objects = map { get_value( $sub => $_ ) } @objects;
        }

        my $func = "_test_".$key;
        if ( defined &{ $func } ) {
            no strict 'refs';
            $func->( $exp => @objects );
            next;
        }

        my @got;
        for my $i ( 0 .. $#objects ) {
            if ( ! ref $expected[ $i ] ) {
                push( @got, get_value( $key => $objects[ $i ] ) );
            } elsif ( ref $expected[ $i ] eq 'ARRAY' ) {
                push( @got, [ get_value( $key => $objects[ $i ] ) ] );
            } elsif ( ref $expected[ $i ] eq 'HASH' ) {
                push( @got, { get_value( $key => $objects[ $i ] ) } );
            } else {
                die "Unknown got value!";
            }
        }

        if ( ! @got ) { die "No values for \@got!" }
        if ( ! @expected ) { die "No values for \@expected!" }

        is_deeply( \@got, \@expected, $sub ? "$sub $key" : $key );
    }
}

sub get_value {
    my ( $key, $object ) = @_;
    if ( blessed $object && $object->can( $key ) ) {
        return $object->$key();
    } else {
        #if ( ref $object ne 'HASH' ) { diag dump $object, $key, caller }
        return $object->{ $key };
    }
}

sub _test_name_order {
    my @order = @{ shift() };
    my @names = map { $_->{ 'name' } } tabs();

    is_deeply( \@names, \@order, "order" );
}

sub _test_selected {
    my $sel = shift;

    my @exp = ( 0 ) x scalar @_;
    $exp[ $sel - 1 ] = 1;
    
    my @got = map { $_->{ 'selected' } } @_;
    is_deeply( \@got, \@exp, "selected" );
}

sub _test_query_form {
    my $query = shift;

    my @got = ();
    my @exp = ( $query ) x scalar @_;
    for my $obj ( @_ ) {
        my @form = $obj->query_form;
        my $tmp = {};
        while ( my ( $key, $value ) = splice( @form, 0, 2 ) ) {
            if ( my $x = $tmp->{ $key } ) {
                $tmp->{ $key } = [ sort
                    ref $x ? @{ $x } : ( $x ),
                    $value,
                ];
            } else {
                $tmp->{ $key } = $value;
            }
        }
        push( @got, $tmp );
    }
    is_deeply( \@got, \@exp, "query_form" );
}

sub get_ok { mech()->get_ok( @_ ) }
sub get {
    my ( $uri, $count )  = @_;
    get_ok( "http://localhost$uri", "Retrieved $uri" );
    ok( my $t = eval content(), 'loaded content' );
    is( @{ $t }, numtabs(), "Correct number of tabs" );
    settabs( $t );
}

sub content { mech()->content }

1;


t/script/testapp_test.pl
#!/usr/bin/env perl

use strict;
use warnings;

use FindBin;
use lib "$FindBin::Bin/..";
use lib "$FindBin::Bin/../../lib";
use Catalyst::Test 'TestApp';

print request($ARGV[0])->content . "\n";

1;


lib/CatalystX/Controller/Tabs.pm
package CatalystX::Controller::Tabs;
use strict;
use warnings;
our $VERSION = '0.02';
our $ID = '$Id: Tabs.pm 11 2008-01-30 15:18:52Z jason $';
use base qw( Catalyst::Controller );
use NEXT;

sub build_tabs : Private {
    my ( $self, $c ) = @_;

    my $conf = $self->config->{ 'tabs' } || {};

    my %tabs = ();
    my $selected;
    my $namespace = $c->action->namespace;
    for my $container ( $c->dispatcher->get_containers( $namespace ) ) {
        for my $action ( values %{ $container->actions } ) {
            my $attrs = $action->attributes;
            my $attr = $attrs->{ 'Tab' } || next;
            $attr = $attr->[ 0 ];
            my ( $label, $order ) = split( /','/, $attr );

            my $tab = $tabs{ $action->name } = {
                name        => $action->name,
                label       => $label,
                selected    => 0,
                order       => $order || 0,
                uri         => $c->uri_for(
                    $c->dispatcher->get_action( $action->name, $namespace ),
                    $c->request->captures,
                    @{ $c->request->arguments },
                    $c->request->query_params,
                ),
            };
        }
    }
    if ( $tabs{ $c->action->name } ) {
        $selected = $c->action->name;
        $tabs{ $selected }->{ 'selected' } = 1;
    }

    my @tabs = sort {
        $a->{ 'order' } <=> $b->{ 'order' }
            ||
        $a->{ 'label' } cmp $b->{ 'label' }
    } values %tabs;

    $c->stash->{ $conf->{ 'stash_key' } || 'tabs' } = \@tabs;
}

1;

__END__

=head1 NAME

CatalystX::Controller::Tabs - Automatically build tab sets for a controller

=head1 SYNOPSIS

  package MyApp::Controller::Something;
  use base qw( Catalyst::Controller CatalystX::Controller::Tabs );
  
  sub friends : Local Tab('Friends') {
    my ( $self, $c ) = @_;
    
    $c->stash->{ 'template' } = 'friends_list.tt2';
  }
  
  sub enemies : Local Tab('Enemies') {
    my ( $self, $c ) = @_;
    
    $c->stash->{ 'template' } = 'enemies_list.tt2';
  }
  
  sub auto : Private {
    my ( $self, $c ) = @_;
    
    $c->forward( 'build_tabs' )
  }

  [%# Template friends_list.tt2 %]
  [% PROCESS tabs.tt2 %]
  <ul>[% FOR enemy IN enemies %]<li>[% enemy %]</li>[% END %]</ul>
  
  [%# Template enemeies_list.tt2 %]
  [% PROCESS tabs.tt2 %]
  <ul>[% FOR friend IN friends %]<li>[% friend %]</li>[% END %]</ul>
  
  [%# Template friends_list.tt2 %]
  <ul>
  [% FOR tab IN tabs %]
      <li[% IF tab.selected %] class="selected">
          <a href="[% tab.uri %]">[% tab.label %]</a>
      </li>
  [% END %]
  </ul>

=head1 DESCRIPTION

This module allows you to add a 'Tab' attribute to action endpoints, and it
will automatically build a data structure suitable for rendering 'tabs' to
switch between the methods that share the same tab structure.

Although this was originally built to help with making tabbed interfaces, it
isn't limited to creating tabs, as it simply collects the information about
the related actions.  Actions are considered to be related if they share a
namespace.

=head1 METHODS

=head2 build_tabs

This is the only method provided by this module, and you need to make sure
you arrange to forward to this method when you want the tab data structure
to be built.

=head1 CONFIGURATION

You can configure this module with the normal L<Catalyst> style of controller
configuration.  The default configuration is:

    __PACKAGE__->config( {
        tabs    => {
            stash_key       => 'tabs',
        }
    } );

=head2 CONFIGURATION OPTIONS

=over

=item stash_key

This configuration option controls the key that will be used to store the
the generated data structure in the stash.

=back

=head1 SAMPLE CSS

Here is some CSS that works with the template included in the synopsis.  It's
probably not exactly what you need, but it should give a decent starting
point...

  ul.tabs {
    text-align: left;
    margin: 1em 0 1em 0;
    border-bottom: 1px solid #6c6;
    list-style-type: none;
    padding: 3px 10px 3px 10px;
  }
      
  ul.tabs li {
    display: inline;
  }
  
  ul.tabs li.selected {
    border-bottom: 1px solid #fff;
    background-color: #fff;
  }
  
  ul.tabs li.selected a {
    background-color: #fff;
    color: #000;
    position: relative;
    top: 1px;
    padding-top: 4px;
  }
  
  ul.tabs li a {
    padding: 3px 4px;
    border: 1px solid #6c6;
    background-color: #cfc;
    color: #666;
    margin-right: 0px;
    text-decoration: none;
    border-bottom: none;
  }
  
  ul.tabs a:hover {
    background: #fff;
}

=head1 SEE ALSO

L<http://www.jasonkohles.com/software/CatalystX-Controller-Tabs>

L<http://catalyst.perl.org/>

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc CatalystX::Controller::Tabs

You can also look for information at:

=over 4

=item * Project home page

L<http://www.jasonkohles.com/software/CatalystX-Controller-Tabs>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/CatalystX-Controller-Tabs>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/CatalystX-Controller-Tabs>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=CatalystX-Controller-Tabs>

=item * Search CPAN

L<http://search.cpan.org/dist/CatalystX-Controller-Tabs>

=back

=head1 BUGS

Please report any bugs or feature requests to C<bug-catalystx-controller-tabs
at rt.cpan.org>, or through the web interface at L<http://rt.cpan.org>.  I
will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head2 KNOWN BUGS

The methods provided for determining the URI for an action don't seem to work
the same way for L<Catalyst::DispatchType::Regex|Catalyst::DispatchType::Regex>
actions, but this hasn't been a big issue for me so far, and I haven't had
time to track the problem down yet (it seems to be an issue with
L<Catalyst::DispatchType::Regex/uri_for_action|uri_for_action in
Catalyst::DispatchType::Regex> if you want to look for it...)

=head1 AUTHOR

Jason Kohles C<< <email@jasonkohles.com> >> L<http://www.jasonkohles.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007-2008 Jason Kohles.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

